nixos/modules/services/misc/nix-daemon.nix

```diff
@@ -192,50 +192,136 @@ in
+              '';
+            };
+            sshUser = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "builder";
+              description = ''
+                The username to log in as on the remote host. This user must be
+                able to log in and run nix commands non-interactively. It must
+                also be privileged to build derivations, so must be included in
+                <option>nix.trustedUsers</option>.
+              '';
+            };
+            sshKey = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "/root/.ssh/id_buildhost_builduser";
+              description = ''
+                The path to the SSH private key with which to authenticate on
+                the build machine. The private key must not have a passphrase.
+                If null, the building user (root on NixOS machines) must have an
+                appropriate ssh configuration to log in non-interactively.
+
+                Note that for security reasons, this path must point to a file
+                in the local filesystem, *not* to the nix store.
+              '';
+            };
+            maxJobs = mkOption {
+              type = types.int;
+              default = 1;
+              description = ''
+                The number of concurrent jobs the build machine supports. The
+                build machine will enforce its own limits, but this allows hydra
+                to schedule better since there is no work-stealing between build
+                machines.
+              '';
+            };
+            speedFactor = mkOption {
+              type = types.int;
+              default = 1;
+              description = ''
+                The relative speed of this builder. This is an arbitrary integer
+                that indicates the speed of this builder, relative to other
+                builders. Higher is faster.
+              '';
+            };
+            mandatoryFeatures = mkOption {
+              type = types.listOf (types.enum [ "big-parallel" "kvm" "benchmark" "nixos-test" ]);
+              default = [];
+              example = [ "big-parallel" ];
+              description = ''
+                A list of features mandatory for this builder. The builder will
+                be ignored for derivations that don't require all features in
+                this list. All mandatory features are automatically included in
+                <varname>supportedFeatures</varname>.
+                Note that there is no fixed set of features. To avoid
+                misconfiguration, this option only accepts a set of typical
+                values.
+                Please use the <varname>mandatoryExtraFeatures</varname> option
+                for mandatory custom features.
+              '';
+            };
+            mandatoryExtraFeatures = mkOption {
```

Comment:
I think in general these options are rather sub-optimally laid out. E.g. currently you do
```nix
{
  buildMachines = [ {
    hostName = "example.com";
    mandatoryFeatures = [ "big-parallel" ];
  } ];
}
```

Which once declared:
- Doesn't allow you to change the config of a machine in other modules since `buildMachines` is of type `listOf`
- Even if that wasn't the case, `mandatoryFeatures` is a list still -> same problem
- In addition, due to `mandatoryFeatures` being a list, it means that one could specify the same settings multiple times, which wouldn't be a problem probably, but it's not nice

So, I'm thinking that this module could use some restructuring in the future to remove any `listOf`'s to allow overriding and such. With this expectation, I think it makes sense to currently only use this PR to make the existing module safer and more documented, such that it doesn't introduce anything other than what's there already. With this in mind, I think these `mandatoryExtraFeatures` and `supportedExtraFeatures` options can be removed, using `listOf str` for the non-Extra variant, falling in line with what it was previously.

Reply:
Works for me. I'll update accordingly.